spectrum-network/src/network_controller.rs

```diff
@@ -171,6 +234,10 @@ where
                         conn_id: *conn_id,
                         enabled_protocols: HashMap::new(),
                     });
+                    // notify all handlers about new connection.
+                    for (_, ph) in self.supported_protocols.values() {
+                        ph.connected(*peer_id);
```

Comment:
Why do we need to notify all protocol handlers?

Reply:
For instance Sync handler should immediately request Sync protocol with new peers.

Reply:
More generally question is - "Why NC should contains THandler instances?". We can just push it swarm with GenerateEvent. Why is it the responsibility of the nc?

Reply:
Why not?

Reply:
I mean, any reasons we should move this outside the swarm, define more events that basically represent the same events coming from PCH + transmit them once more?

Reply:
The libp2p architecture offers you to transfer all the protocol messages through the swarm. But you added new different queues for this purpose. Explain why you choose not to use the native solution but create your own? It makes logic more complicated I think

Reply:
Yes, libp2p doesn't tell us, how EXECTLY we should implement this, but here is an example how developers would like us to use it. Take a look on NetworkBehaviour derivation - it gives us an opportunity to transmit custom messages through swarm and poll it from it. Substrate implementation also follows this approach 

Examples: 
https://github.com/libp2p/rust-libp2p/tree/master/examples
https://github.com/paritytech/substrate/blob/e6768a3bd553ddbed12fe1a0e4a2ef8d4f8fdf52/client/network/src/protocol/notifications/behaviour.rs#L1957

Reply:
Substrate only propagates custom protocols notifications upwards. It has to do so in order to provide its users with a possibility to define custom behaviours outside the networking crate which is not our case. Also, I don't think substrate networking is a good reference in terms of architecture, it's pretty messy.

Reply:
If you are looking for a "correct" way of handling messages take a look at protocols provided by libp2p, e.g https://github.com/libp2p/rust-libp2p/blob/master/protocols/gossipsub/src/behaviour.rs

Reply:
Don't ignore "Take a look on NetworkBehaviour derivation - it gives us an opportunity to transmit custom messages through swarm and poll it from it.", please.

About gossibsub. Check https://github.com/libp2p/rust-libp2p/blob/c32f03c3170a8123cb480ece9aa511a6f9ea6dd5/protocols/gossipsub/src/behaviour.rs#L3441

Reply:
Right, it propagates domain events outside the swarm which is not our case, while internal messages are handled by behaviour which is exactly its responsibility according to docs. 
So the lib provides us with a way to propagate any event through the swarm. If there are no more reasons why we should stick with further propagation of protocol messages in our application, then I'm resolving the convo.

Reply:
Why it is not our case? It's your decision to create another way of processing protocol events. Can you please explain for what purpose, please?

Reply:
Libp2p provides none ways of processing protocol messages as we've found out. What's wrong with the approach I use in this PR? If we decided to pass messages through swarm we would simply have to move routing logic from behavior to something that wraps swarm. What's the point? Or am I missing something?

Reply:
About docs (They are not approved).

"Network Controller (NC) syncronises behaviours of libp2p swarm, PCH for each connection and PM. In order to do that NC 
has to track a projection of states of enabled protocols (only the relevant information, thus this set of states is 
more narrow than the one in PCH)"

No info about NC protocol event processing.

Reply:
I meant libp2p docs:
"[NetworkBehaviour](https://docs.rs/libp2p/latest/libp2p/swarm/trait.NetworkBehaviour.html) defines what bytes to send and to whom."

Reply:
The problem is that you're trying to create different way of processing business event without any reason. You're trying to mix network logic (libp2p entities) and business logic (THandler's)

Reply:
Different from what? Have you seen an alternative mechanism for processing?
Secondly, please show where network logic is mixed with business logic.

Reply:
All examples shows you, that business logic should poll swarm -> do some stuff -> tell swarm what entity should be sent. You're ignoring this approach, why?

Reply:
Those examples showcase some trivial applications built on top of some protocol shipped with libp2p (e.g. Gossipsub). That's why business events are emitted via swarm.
In our case we have N protocols which should be ran concurrently. If we choose to use swarm as a gateway for all events from all protocols we would introduce a single point of synchronization.

Reply:
@Bromel777 I'd like to hear some constructive reasoning from you - e.g. what consequences a particular design solution implies in application to our use case.

Reply:
Yep, we have N protocols, but there is no difference from examples. Check substrate.
Inject_event is a single point of synchronization in your implementation.
About, libp2p mixing entities should only be responsible for network part. In you're case NC (i.e libp2p behavior) additionally invoke protocols logic.

Reply:
You should look carefully at how messages are sent to peers in spectrum and in substrate. Spoiler: in both cases they are enqueued directly to PCHs buffers.

Reply:
On mixing. It seems to me that NC only injects a fixed set of network events to prot handlers. While business logic of concrete protocols is defined in ProtocolBehaviour implementors.